### ISSUE TITLE
Define format for dynamic messages

### ISSUE THREAD

**Issue by ebruchez** [orbeon/orbeon-forms#3078]
We have a number of issues which call for a template format:

- fr:explanation to support dynamic content #2427
- FB: Dynamic labels, hints and error messages #1342
- Email to support dynamic content #2428

They all should use the same format. Pasting below a comment which was in #1342:

After discussion with @avernet, thinking that there are options to handle the template:

- XPath value templates
- Java message format templates
- maybe others (just using `$foo` in the string as in Scala string interpolation?)

Since we support internationalization, we need to ensure that things are consistent

- the parameters to the template should be fixed and shared between languages
- on the other hand a template in a given language can omit some or all of the parameters

So it makes sense to have the ability to specify parameters separately in a list, which probably calls for doing this in Control Settings (and Section Settings, and Form Settings if we handle the title).

How would using a list apply to email templates?

**Comment by ebruchez** [comment]
This moved from #2427:

As an example of what we use for resources:

```
Your form {
    1,
    choice,
    0#does not contain any messages
    |
    1#contains the following '{
        0,
        choice,
        0#error|1#warning|2#informational message|3#message
    }'
    |
    1<contains the following '{
        0,
        choice,
        0#errors|1#warnings|2#informational messages|3#messages
    }'
}
```

Here, instead of positional parameters (`0`, `1`, etc.) we would use variables:

```
Your form {
    $count,
    choice,
    0#does not contain any messages
    |
    1#contains the following '{
        $type,
        choice,
        0#error|1#warning|2#informational message|3#message
    }'
    |
    1<contains the following '{
        $type,
        choice,
        0#errors|1#warnings|2#informational messages|3#messages
    }'
}
```

where `$count` and `$type` would refer to other controls.

**Comment by bitumnus** [comment]
wait for this soooo much.... Need some help maybe??

**Comment by avernet** [comment]
@bitumnus As a first step, we need to decide on the format of the template and find a format we find simple enough to use for typical users of Form Builder, yet powerful enough, and that we can use consistently in several places (email, placeholder control, labels, help, hint, alert, and maybe more). You'll let use know if you have any thoughts on this.

**Comment by bitumnus** [comment]
@avernet  Clear. 
I read earlier that need the ability to send messages to multiple people #2492 . I was not looking for a solution (perhaps it already is), but I do this very well. In builder, when adding a new text field, in Control Settings (Basic) mark checkbox "Email Recipient" in every field we'll need. In form Runner write in these fields e-mail addresses (one by one), and voila. All I have done today and tested. Maybe it help someone... Or maybe I don't understood what's problem )) (English is not my native language)
And about this item, can be on the same principle: add in 'Settings Control' checkbox e.x. "attach to the body." and transmit the value of this field (like in recipient property email.to). And now I understand that 'body' not a property... I'll think about it. 
Because, really, it should be a powerful tool not only for email, like you say.

**Comment by avernet** [comment]
@bitumnus You're right: you can already send an email to multiple people. However, to clarify, #2492 is calling for supporting the ability to send *different* emails to different people, like a confirmation email to person who did the submission, and a different message to a person who will be processing the request.

**Comment by ebruchez** [comment]
So I think that Java's `MessageFormat` is the best so far, at least as the format which we store in the code, and probably as the one we show in the UI: it handles text, numbers, and choices, which are very useful. And it looks more like a resource than code.

XPath value templates could work but you would have to put expressions in `{...}` in all cases, which makes things harder. Also, internationalization would then become harder.

So voting for `MessageFormat` for now. See a UI mockup in #1342.

**Comment by bitumnus** [comment]
@ebruchez  perfectly well! I like this sentence about Java's format! 
now, I coped with the insertion of a link to the e-mail, which is sent from form.  download email-form.xpl file  from this source [https://github.com/orbeon/orbeon-forms/blob/tag-release-4.3.1.0-ce/src/resources/apps/fr/email/email-form.xpl#L224](url) and put in necessary parameters in email body (after dedicated line), like: 
```
<xsl:variable name="document" select="doc('input:parameters')/*/document" as="xs:string"/>
 <part name="text" content-type="text/plain">
                         <!-- <xsl:value-of select="concat($fr-resources/resource[@xml:lang = $request-language]/email/body,'')"/>  -->
			<xsl:value-of select="$fr-resources/resource[@xml:lang = $request-language]/email/body"/>
			<xsl:value-of select="concat(' ', ' ')"/>
                    	<xsl:value-of select="p:rewrite-service-uri(concat('/fr/',$app,'/',$form, '/view/',$document), true())"/>
```
and in property-local.xml in` fr.resource.*.*email.body` write static text. And now when sent email, it contain text + link to form. 
Now it is the best I came up with. Because the solution had to be "yesterday". Perhaps someone it will help, but really want to see the option easier.

**Comment by ebruchez** [comment]
@bitumnus Good, thanks for sharing. Hopefully we will get to implement this soon so that kind of workarounds are no longer needed.

**Comment by ebruchez** [comment]
We are now thinking again about the format. The Java format is weird and not that powerful. Its main benefits ar:

- declarative
- support of choices

But the format is not quite ideal in that, for example, types and formats to not match the types and formats we use. Further, we would like named parameters.

So we are thinking about using a similar-but-different syntax. For plain values with default formatting based on the runtime types of the provided values:

```
Hi {name}, you have {count} file(s).
```

Or:

```
Hi {$name}, you have {$count} file(s).
```

The tricky part is the handling of choices if possible in a declarative way. It is not a number one priority, but a really nice to have.

**Comment by ebruchez** [comment]
If the content of the brackets is an expression, it could be initially limited to referring to variables. If an expression itself contains nested templates, for example as the result of a choice selector, there has to be a way to parse that into something like an AVT, ideally.

**Comment by ebruchez** [comment]
Something like this could be envisioned:

```
This is a template with a number value {$number}. This is a choice of {
 switch ($whatever) {
  case 0 =>
    'some string'
  case 1 =>
    'some other string'
  case value if $value gt 1 =>
    'some thing with a nested template {$value} or {$whatever}'.
 }
}
```

The constructs within brackets would be limited to:

1. variable references using `$`
2. switch constructs with a very specific, limited matching syntax

Nested strings are also templates, recursively.

The template could be rewritten into a standard XPath expression.

**Comment by ebruchez** [comment]
If we wanted to use pure XPath:

```
This is a template with a number value {$number}. This is a choice of {
  if ($whatever = 0) then
    'some string'
  else if ($whatever = 1) then
    'some other string'
  else
    xxf:template('some thing with a nested template {{$value}} or {{$whatever}}')
}
```

**Comment by ebruchez** [comment]
In XQuery, I think you could write something like:

```xquery
<dummy>
  This is a template with a number value {$number}. This is a choice of {
    if ($whatever = 0) then
      'some string'
    else if ($whatever = 1) then
      'some other string'
    else
      <dummy>some thing with a nested template {$value} or {$whatever}</dummy>
  }
</dummy>
```

**Comment by ebruchez** [comment]
In Scala-like syntax:

```
s"""
  This is a template with a number value {$number}. This is a choice of {
    if ($whatever = 0) then
      'some string'
    else if ($whatever = 1) then
      'some other string'
    else
      s"some thing with a nested template {$value} or {$whatever}"
  }
"""
```

**Comment by ebruchez** [comment]
At first, we won't support nested expressions, but only variable references.

**Comment by ebruchez** [comment]
So we now have a syntax for simple parameters, but we are still considering how to do choices. Do we want to follow the last XPath-based syntax proposed above?

We would like the template declaration to remain declarative. This means support switching is only based on comparing a given parameter to a given value (equality, smaller, larger). Using `switch/case`, `match/case`, or `if/then/else` doesn't make much of a difference there as long as the syntax is clear. `case`s would maybe need a specific syntax like `case < 2 =>`, for example.

**Comment by ebruchez** [comment]
Example:

```
This is a template with a number value {$number}. This is a choice of {
  $whatever match {
    case 0   => 'some string'
    case 1   => 'some other string'
    case > 1 => 'some thing with a nested template {$value} or {$whatever}'.
  }
}
```

**Comment by ebruchez** [comment]
The UI could allow editing templates with choices.

**Comment by ebruchez** [comment]
As of Orbeon Forms 2018.1, we have a basic template format which supports plain variables.

**Comment by ebruchez** [comment]
So what's left:

- [ ] format for conditions

**Comment by avernet** [comment]
Resolution: Closing, as we are currently satisfied with the basic template format that supports plain variables. We can always create another issue in the future if we want more.
